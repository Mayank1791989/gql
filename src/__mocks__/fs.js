/* @flow */
import MemoryFileSystem from 'memory-fs';
import path from 'path';

function createNewMemoryFileSystem(): MemoryFileSystem {
  const _fs = new MemoryFileSystem();
  // Note: binding all methods of _fs
  // some libs use individual fs method
  // e.g const writeFileSync = fs.writeFileSync
  //  writeFileSync() // this will call method with wrong this
  // happening in find-config package
  Object.keys(MemoryFileSystem.prototype).forEach((key) => {
    if (typeof _fs[key] === 'function') {
      _fs[key] = _fs[key].bind(_fs);
    }
  });
  return _fs;
}

const fs = createNewMemoryFileSystem();

// This is a custom function that our tests can use during setup to create
// files in memory
function __setMockFiles(newMockFiles) {
  Object.keys(newMockFiles).forEach((file) => {
    const dirname = path.dirname(file);
    fs.mkdirpSync(dirname);
    fs.writeFileSync(file, newMockFiles[file], 'utf8');
  });
}

fs.__setMockFiles = __setMockFiles;

module.exports = fs;
