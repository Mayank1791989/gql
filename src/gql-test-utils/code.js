/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
