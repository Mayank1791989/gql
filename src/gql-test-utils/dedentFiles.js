/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
