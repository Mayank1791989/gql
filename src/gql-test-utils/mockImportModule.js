/* @flow */
export default function mockImportModule(
  moduleName: string,
  mockFactory: () => $FixMe,
) {
  jest.doMock('import-from', () => {
    return function importFrom(dir, moduleID) {
      if (moduleID === moduleName) {
        return mockFactory();
      }
      const importFromActual = jest.requireActual('import-from');
      return importFromActual(dir, moduleID);
    };
  });
}
