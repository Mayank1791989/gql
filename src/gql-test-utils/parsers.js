/* @flow */
import SchemaParser from 'gql-schema-parser-default';
import EmbeddedQueryParser from 'gql-query-parser-embedded-queries';
import QueryParser from 'gql-query-parser-default';
import QueryPresetRelay from 'gql-query-preset-relay';

export function relayQLParser() {
  return new EmbeddedQueryParser({
    start: 'Relay\\.QL`',
    end: '`',
    allowFragmentWithoutName: true,
    allowFragmentInterpolation: true,
  });
}

export function apolloParser() {
  return new EmbeddedQueryParser({
    start: 'gql`',
    end: '`',
    allowDocumentInterpolation: true,
  });
}

export const relayExtendSchema = QueryPresetRelay().extendSchema;

export function queryParser() {
  return new QueryParser();
}

export function customParser() {
  return new EmbeddedQueryParser({
    start: '"""',
    end: '"""',
  });
}

export function embeddedQueryParser({
  start,
  end,
}: {
  start: string,
  end: string,
}) {
  return new EmbeddedQueryParser({
    start,
    end,
  });
}

export function schemaParser() {
  return new SchemaParser();
}
