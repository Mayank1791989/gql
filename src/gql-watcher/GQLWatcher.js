/* @flow */
import boolValue from 'gql-shared/boolValue';
import log from 'gql-shared/log';
import sane from 'sane';
import os from 'os';
import { execSync } from 'child_process';

import watch, { type WatchOptions } from './watch';
import { memoizeSingle } from 'gql-shared/memoize';
import { type IWatcher } from './types';

const logger = log.getLogger('gql-watcher');

type Options = $ReadOnly<{|
  watchman?: boolean,
  watch?: boolean,
|}>;

export default class GQLWatcher {
  _watchers: Array<*> = [];
  _options: *;

  constructor(options: Options) {
    this._options = options;
  }

  watch(options: $Rest<WatchOptions, {| Watcher: any, watch: any |}>) {
    logger.info(`setting watcher for files ${JSON.stringify(options.files)}`);
    const watcher = watch({
      ...options,
      Watcher: this._getWatcher(boolValue(this._options.watchman, true)),
      watch: boolValue(this._options.watch, true),
    });
    this._watchers.push(watcher);
    return watcher;
  }

  close() {
    this._watchers.forEach(watcher => {
      watcher.close();
    });
  }

  _getWatcher = memoizeSingle((useWatchmanIfAvailable: boolean): Class<
    IWatcher,
  > => {
    if (useWatchmanIfAvailable && this._checkWatchmanInstalled()) {
      logger.info('using "watchman" watcher.');
      return sane.WatchmanWatcher;
    }

    if (os.platform() === 'darwin') {
      logger.info('using "fsevents" watcher.');
      return sane.FSEventWatcher;
    }

    logger.info('using "node" watcher.');
    return sane.NodeWatcher;
  });

  _checkWatchmanInstalled() {
    logger.info('checking "watchman" version...');
    try {
      const version = execSync('watchman --version', {
        stdio: ['ignore'],
        encoding: 'utf8',
      });
      logger.info(`watchman v${version.toString().trim()} found.`);
      return true;
    } catch (e) {
      logger.info('"watchman" not found.');
      return false;
    }
  }
}
