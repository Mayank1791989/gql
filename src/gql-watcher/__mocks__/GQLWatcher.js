/* @flow */
const GQLWatcher = require.requireActual('../GQLWatcher').default;

export default class GQLWatcherMock extends GQLWatcher {
  constructor(options) {
    options.watchman = false;
    super(options);
  }
}
