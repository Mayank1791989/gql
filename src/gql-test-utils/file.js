/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import fs from 'fs-extra';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    fs.outputFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  fs.removeSync(TEST_TEMP_DIR);
}
