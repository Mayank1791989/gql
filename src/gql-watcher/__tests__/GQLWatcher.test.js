/* @flow */
/* eslint-disable no-sync */
import GQLWatcher from '../GQLWatcher';
import { createTempFiles } from 'gql-test-utils/file';
import path from 'path';
import fs from 'fs-extra';

async function setupWatcher(onChange: Function) {
  const rootPath = createTempFiles({
    // Putting Date.now() to create diff directory every time test run
    // we can also run tests in parallel.
    '.git/tmp': `to trigger watchman events. ${Date.now()}`,
    'watchdir/dirA/fileA': 'fileA',
    'watchdir/dirA/fileB.js': 'fileB.js',
    'watchdir/dirA/fileC.js': 'fileC.js',
    'watchdir/dirA/fileD.js': 'fileD.js',
    'outsidewatchdir/dirA/fileA.js': 'fileA.js',
  });

  const watcher = new GQLWatcher({ watch: true });
  const watchSubscription = watcher.watch({
    rootPath,
    files: 'watchdir/**/*.js',
    onChange: files => {
      const sortedFiles = files.sort((a, b) => {
        if (a.name > b.name) {
          return 1;
        }
        return -1;
      });
      onChange(sortedFiles);
    },
  });

  await watchSubscription.onReady();
  return { rootPath, watcher };
}

describe('add', () => {
  test('add file triggers add event', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);
    // create file
    await fs.outputFile(
      path.join(rootPath, 'watchdir/dirA/fileX.js'),
      'fileX.js',
    );
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('add file in new dir triggers add event', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);
    // create file
    await fs.outputFile(
      path.join(rootPath, 'watchdir/dirB/fileA.js'),
      'fileA.js',
    );
    // wait for all events called
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('add file outside watch root should not trigger add event', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    // create file
    await fs.outputFile(
      path.join(rootPath, 'outsidewatchdir/dirA/fileX.js'),
      'fileX.js',
    );

    // wait for all events called
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('add dir should not trigger add event', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    // create file
    await fs.mkdirp(path.join(rootPath, 'watchdir/dirB'));

    // wait for all fs events called
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });
});

describe('delete', () => {
  test('file delete trigger delete event on deleted file', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.remove(path.join(rootPath, 'watchdir/dirA/fileB.js'));
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('directory delete should trigger delete event on all files inside deleted dir', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.remove(path.join(rootPath, 'watchdir/dirA'));
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('directory delete outside root should not trigger event', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.remove(path.join(rootPath, 'outsidewatchdir/dirA'));
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });
});

describe('move (rename)', () => {
  test('rename file should trigger two events delete and add', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.move(
      path.join(rootPath, 'watchdir/dirA/fileB.js'),
      path.join(rootPath, 'watchdir/dirA/fileBB.js'),
    );
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test('move file from one dir to other dir should trigger two events delete and add', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.move(
      path.join(rootPath, 'watchdir/dirA/fileB.js'),
      path.join(rootPath, 'watchdir/dirB/fileB.js'),
    );
    await waitForFsEvents();

    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  test.only('rename dir should trigger two events delete and add for all files inside dir', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    await fs.move(
      path.join(rootPath, 'watchdir/dirA'),
      path.join(rootPath, 'watchdir/dirC'),
    );
    await waitForFsEvents();

    // console.log(JSON.stringify(onChangeMock.mock.calls, null, 2));
    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });

  // When sync is used fs events emitted are different
  test('[using sync] rename dir should trigger two events delete and add for all files inside dir', async () => {
    const onChangeMock = jest.fn();
    const { rootPath, watcher } = await setupWatcher(onChangeMock);

    fs.moveSync(
      path.join(rootPath, 'watchdir/dirA'),
      path.join(rootPath, 'watchdir/dirC'),
    );
    await waitForFsEvents();

    // console.log(JSON.stringify(onChangeMock.mock.calls, null, 2));
    expect(onChangeMock.mock.calls).toMatchSnapshot();

    await watcher.close();
  });
});

function waitForFsEvents(): Promise<void> {
  return wait(1000);
}

function wait(timeInMsec: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, timeInMsec);
  });
}
