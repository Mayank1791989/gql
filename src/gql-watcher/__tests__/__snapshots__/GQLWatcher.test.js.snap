// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add add dir should not trigger add event 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
]
`;

exports[`add add file in new dir triggers add event 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirB/fileA.js",
      },
    ],
  ],
]
`;

exports[`add add file outside watch root should not trigger add event 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
]
`;

exports[`add add file triggers add event 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileX.js",
      },
    ],
  ],
]
`;

exports[`delete directory delete outside root should not trigger event 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
]
`;

exports[`delete directory delete should trigger delete event on all files inside deleted dir 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
]
`;

exports[`delete file delete trigger delete event on deleted file 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
    ],
  ],
]
`;

exports[`move (rename) [using sync] rename dir should trigger two events delete and add for all files inside dir 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileD.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileD.js",
      },
    ],
  ],
]
`;

exports[`move (rename) move file from one dir to other dir should trigger two events delete and add 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirB/fileB.js",
      },
    ],
  ],
]
`;

exports[`move (rename) rename dir should trigger two events delete and add for all files inside dir 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileD.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirC/fileD.js",
      },
    ],
  ],
]
`;

exports[`move (rename) rename file should trigger two events delete and add 1`] = `
Array [
  Array [
    Array [
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileC.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileD.js",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "exists": false,
        "name": "watchdir/dirA/fileB.js",
      },
      Object {
        "exists": true,
        "name": "watchdir/dirA/fileBB.js",
      },
    ],
  ],
]
`;
